url1f1de5fc71_cooijmanstim_hobo_tgz-pJ8-hobo_TestingEnviromentJ8/src/hobo/ParameterTuning.java

```diff
@@ -0,0 +1,164 @@
+				Game g = new Game("verbose:false",
+				                  Player.fromConfiguration("uncertain minimax    verbose:false sample_size:3 decision_time:3000"+parameters),
+				                  Player.fromConfiguration("uncertain montecarlo verbose:false sample_size:3 decision_time:3000"));
+				g.play();
+				return 1.0 * Math.signum(g.state.aheadness(0));
+			}
+		});
+	}
+
+	public static void tuneBelief() {
+		System.out.println("tuning belief");
+		tuneByCrossEntropy(5, 2, 3, new double[]{ 2 }, new double[]{ 1 }, new Function<double[], Double>() {
+			@Override public Double call(double[] xs) {
+				String parameters = " belief_alpha:"+xs[0];
+				Game g = new Game("verbose:false",
+				                  Player.fromConfiguration("uncertain minimax    verbose:false sample_size:3 decision_time:6000"+parameters),
+				                  Player.fromConfiguration("uncertain montecarlo verbose:false sample_size:3 decision_time:6000"+parameters));
+				System.out.println(xs[0]);
+				g.play();
+				
+				double[] accuracies = new double[g.players.length];
+				for (int i = 0; i < g.players.length; i++) {
+					double[] stats = g.players[i].statistics();
+					accuracies[i] = stats[stats.length-1];
+				}
+				return Util.mean(accuracies);
+			}
+		});
+	}
+
+	private static final int nthreads = 2;
+	public static void tuneByCrossEntropy(int population_size, int selection_size, int sample_size,
+		                                  double[] initial_means, double[] initial_stdevs,
+		                                  final Function<double[],Double> evaluation) {
+		final double alpha = 0.1;
+		double[] means = initial_means.clone(), stdevs = initial_stdevs.clone();
+		final int nvariables = means.length;
+
+		final double[][] population = new double[population_size][nvariables];
+		final long[] failures_and_tries = new long[2];
+
+		ExecutorService pool = Executors.newFixedThreadPool(nthreads);
+		
+		while (true) {
+			System.out.println("means:  "+Arrays.toString(means));
+			System.out.println("stdevs: "+Arrays.toString(stdevs));
+
+			for (int i = 0; i < population_size; i++) {
+				for (int j = 0; j < nvariables; j++)
+					population[i][j] = means[j] + random.nextGaussian() * stdevs[j];
+			}
+
+			// fuck you java
+			Object[] future_evaluations = new Object[population_size];
+			for (int i = 0; i < population_size; i++) {
+				for (int k = 0; k < sample_size; k++) {
+					final int j = i;
+					future_evaluations[i] = pool.submit(new Callable<Double>() {
+						@Override public Double call() {
+							return evaluation.call(population[j]);
+						}
+					});
+				}
+			}
+
+			final double[] evaluations = new double[population_size];
+			for (int i = 0; i < population_size; i++) {
+				for (int k = 0; k < sample_size; k++) {
+					try {
+						failures_and_tries[1]++;
+
+						while (true) {
+							try {
+								evaluations[i] = ((Future<Double>)future_evaluations[i]).get();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:** 
The loop containing this operation may iterate indefinitely if the operation throws an exception on multiple instances of its evaluation. As the operation throws an exception, it bypasses exit points in the loop and continues to the next iteration through a catch clause.

**Fix:** 
Add a retry counter to ensure that the loop terminates after a predefined number of iterations.